File: src/grids/vertical.jl
"""
Vertical grid family (1D column).

Implements the public API contract defined in `docs/GRIDS_API.md` §2.3, §3.1
and §7 for the Julia binding. The generator entry point lives in the
`EarthSciDiscretizations.grids` submodule (wired at the bottom of
`src/EarthSciDiscretizations.jl`) and is invoked as:

    using EarthSciDiscretizations
    g = EarthSciDiscretizations.grids.vertical(; coordinate = :sigma, nz = 16)
    g = EarthSciDiscretizations.grids.vertical(; coordinate = :z,
                                                 levels = [0.0, 500.0, 1500.0])
    g = EarthSciDiscretizations.grids.vertical(;
            coordinate = :eta, ak = ak_vec, bk = bk_vec, p0 = 1.0e5)

Supported coordinate kinds (rhymes with the Python sibling,
`python/src/earthsci_toolkit/grids/vertical.py`):

- `:sigma` — terrain-following, dimensionless ∈ [0, 1]; 1 at surface, 0 at top.
- `:eta` — hybrid sigma-pressure (NCAR / CAM style). Synthesized sigma
  = `ak/p0 + bk` at interfaces; must be strictly decreasing.
- `:z` — geometric altitude (m); strictly increasing.
- `:theta` — potential temperature (K); strictly increasing.
- `:hybrid_sigma_theta` — blended sigma → theta; sigma-like levels.
- `:z_star` — generalized geometric height; strictly increasing.

Per the 2026-04-20 scope correction on the vertical fixtures bead, the
`.esm` lowering is a small declarative config (family + coordinate kind +
interface levels + hybrid coefficients when applicable + provenance), NOT a
serialized geometry blob. Cell centers and widths are derived on demand
from `levels` via pure arithmetic.
"""

# ---------------------------------------------------------------------------
# Type
# ---------------------------------------------------------------------------

struct VerticalGrid{T <: AbstractFloat} <: AbstractGrid
    coordinate::Symbol
    levels::Vector{T}
    centers::Vector{T}
    widths::Vector{T}
    ak::Vector{T}
    bk::Vector{T}
    p0::T
    dtype::String
    ghosts::Int
end

Base.ndims(::VerticalGrid) = 1
Base.eltype(::VerticalGrid{T}) where {T} = T

n_cells(g::VerticalGrid) = length(g.levels) - 1
n_vertices(g::VerticalGrid) = length(g.levels)
n_edges(g::VerticalGrid) = length(g.levels) - 1

const _VERTICAL_FAMILY_VERSION = "1.0.0"

const _VERTICAL_COORDINATES =
    (:sigma, :eta, :z, :theta, :hybrid_sigma_theta, :z_star)

const _VERTICAL_METRIC_NAMES = (:dz, :z, :sigma, :pressure, :ak, :bk)

# ---------------------------------------------------------------------------
# Parse + validation helpers
# ---------------------------------------------------------------------------

_parse_vertical_coordinate(c::Symbol) =
    c in _VERTICAL_COORDINATES ? c :
        throw(ArgumentError(
            "vertical: unknown coordinate :$c; expected one of " *
                "$(_VERTICAL_COORDINATES)"
        ))
_parse_vertical_coordinate(c::AbstractString) =
    _parse_vertical_coordinate(Symbol(c))

function _vertical_uniform_sigma(nz::Int, ::Type{T}) where {T}
    nz ≥ 1 || throw(ArgumentError("vertical: nz must be ≥ 1; got $nz"))
    return T[one(T) - T(k) / T(nz) for k in 0:nz]
end

function _vertical_coerce_levels(
        levels, ::Type{T};
        must_decrease::Bool,
        domain::Union{Tuple{Real, Real}, Nothing} = nothing,
        label::AbstractString = "levels"
    ) where {T}
    arr = T[T(x) for x in levels]
    length(arr) ≥ 2 ||
        throw(ArgumentError("vertical: `$label` must have ≥ 2 entries; got $(length(arr))"))
    for x in arr
        isfinite(x) ||
            throw(DomainError(x, "vertical: `$label` must be finite"))
    end
    if domain !== nothing
        lo, hi = T(domain[1]), T(domain[2])
        for x in arr
            (x < lo || x > hi) &&
                throw(DomainError(x,
                    "vertical: `$label` entries must lie in [$lo, $hi]"))
        end
    end
    if must_decrease
        for k in 1:(length(arr) - 1)
            arr[k + 1] < arr[k] ||
                throw(DomainError(arr,
                    "vertical: `$label` must be strictly decreasing"))
        end
    else
        for k in 1:(length(arr) - 1)
            arr[k + 1] > arr[k] ||
                throw(DomainError(arr,
                    "vertical: `$label` must be strictly increasing"))
        end
    end
    return arr
end

function _vertical_coerce_hybrid(
        coeffs, ::Type{T}, expected_len::Int, label::AbstractString
    ) where {T}
    coeffs === nothing &&
        throw(ArgumentError("vertical: `$label` is required"))
    arr = T[T(x) for x in coeffs]
    length(arr) == expected_len ||
        throw(ArgumentError(
            "vertical: `$label` must have length nz+1 = $expected_len; " *
                "got $(length(arr))"
        ))
    for x in arr
        isfinite(x) ||
            throw(DomainError(x, "vertical: `$label` must be finite"))
    end
    return arr
end

function _vertical_centers_and_widths(levels::Vector{T}) where {T}
    n = length(levels) - 1
    centers = Vector{T}(undef, n)
    widths = Vector{T}(undef, n)
    for k in 1:n
        centers[k] = (levels[k] + levels[k + 1]) / T(2)
        widths[k] = abs(levels[k + 1] - levels[k])
    end
    return centers, widths
end

# ---------------------------------------------------------------------------
# Generator
# ---------------------------------------------------------------------------

"""
    _vertical(; coordinate, nz=nothing, levels=nothing,
                ak=nothing, bk=nothing, p0=1.0e5, transition=nothing,
                dtype=Float64, ghosts=0) -> VerticalGrid{T}

Primary constructor for the vertical family. See `docs/GRIDS_API.md` §2.3 for
the cross-binding signature contract. Public entry point (Julia):
`EarthSciDiscretizations.grids.vertical`.
"""
function _vertical(;
        coordinate = nothing,
        nz::Union{Int, Nothing} = nothing,
        levels = nothing,
        ak = nothing,
        bk = nothing,
        p0::Real = 1.0e5,
        transition::Union{Real, Nothing} = nothing,
        dtype::Type{T} = Float64,
        ghosts::Int = 0
    ) where {T <: AbstractFloat}

    coordinate === nothing &&
        throw(ArgumentError("vertical: required option `coordinate` is missing"))
    coord = _parse_vertical_coordinate(coordinate)

    ghosts ≥ 0 || throw(ArgumentError("vertical: ghosts must be ≥ 0; got $ghosts"))

    if nz !== nothing
        nz ≥ 1 || throw(ArgumentError("vertical: nz must be ≥ 1; got $nz"))
    end

    p0_T = T(p0)
    (p0_T > zero(T) && isfinite(p0_T)) ||
        throw(DomainError(p0, "vertical: p0 must be positive and finite"))

    local lv::Vector{T}
    local ak_arr::Vector{T}
    local bk_arr::Vector{T}

    if coord === :sigma
        if levels !== nothing
            lv = _vertical_coerce_levels(
                levels, T; must_decrease = true, domain = (0.0, 1.0)
            )
            nz !== nothing && length(lv) != nz + 1 &&
                throw(ArgumentError(
                    "vertical: sigma: nz=$nz inconsistent with levels length $(length(lv))"
                ))
        else
            nz === nothing &&
                throw(ArgumentError("vertical: sigma requires `nz` or `levels`"))
            lv = _vertical_uniform_sigma(nz, T)
        end
        ak_arr = T[]
        bk_arr = T[]

    elseif coord === :z || coord === :theta || coord === :z_star
        levels === nothing &&
            throw(ArgumentError("vertical: :$coord requires explicit `levels`"))
        lv = _vertical_coerce_levels(levels, T; must_decrease = false)
        nz !== nothing && length(lv) != nz + 1 &&
            throw(ArgumentError(
                "vertical: $coord: nz=$nz inconsistent with levels length $(length(lv))"
            ))
        ak_arr = T[]
        bk_arr = T[]

    elseif coord === :eta
        ak === nothing && throw(ArgumentError("vertical: :eta requires `ak` (length nz+1)"))
        bk === nothing && throw(ArgumentError("vertical: :eta requires `bk` (length nz+1)"))
        ak_probe = T[T(x) for x in ak]
        bk_probe = T[T(x) for x in bk]
        length(ak_probe) == length(bk_probe) ||
            throw(ArgumentError(
                "vertical: :eta ak/bk must have equal length; " *
                    "got $(length(ak_probe)) vs $(length(bk_probe))"
            ))
        nz_eff = nz === nothing ? length(ak_probe) - 1 : nz
        nz_eff ≥ 1 || throw(ArgumentError("vertical: nz must be ≥ 1; got $nz_eff"))
        ak_arr = _vertical_coerce_hybrid(ak, T, nz_eff + 1, "ak")
        bk_arr = _vertical_coerce_hybrid(bk, T, nz_eff + 1, "bk")
        sigma = ak_arr ./ p0_T .+ bk_arr
        for k in 1:(length(sigma) - 1)
            sigma[k + 1] < sigma[k] ||
                throw(DomainError(sigma,
                    "vertical: :eta synthesized sigma (ak/p0 + bk) must be " *
                        "strictly decreasing"
                ))
        end
        lv = sigma

    elseif coord === :hybrid_sigma_theta
        if levels === nothing && nz === nothing
            throw(ArgumentError(
                "vertical: :hybrid_sigma_theta requires `nz` or `levels`"
            ))
        end
        if levels !== nothing
            lv = _vertical_coerce_levels(
                levels, T; must_decrease = true, domain = (0.0, 1.0)
            )
            nz !== nothing && length(lv) != nz + 1 &&
                throw(ArgumentError(
                    "vertical: hybrid_sigma_theta: nz=$nz inconsistent with " *
                        "levels length $(length(lv))"
                ))
        else
            lv = _vertical_uniform_sigma(nz, T)
        end
        if transition !== nothing
            (zero(T) < T(transition) < one(T)) ||
                throw(DomainError(transition,
                    "vertical: hybrid_sigma_theta `transition` must be in (0, 1)"
                ))
        end
        ak_arr = ak === nothing ? T[] :
            _vertical_coerce_hybrid(ak, T, length(lv), "ak")
        bk_arr = bk === nothing ? T[] :
            _vertical_coerce_hybrid(bk, T, length(lv), "bk")

    else
        # Unreachable: _parse_vertical_coordinate already validated the set.
        throw(AssertionError("vertical: unreachable coordinate :$coord"))
    end

    centers, widths = _vertical_centers_and_widths(lv)
    return VerticalGrid{T}(
        coord, lv, centers, widths, ak_arr, bk_arr, p0_T,
        _dtype_string(T), ghosts,
    )
end

# ---------------------------------------------------------------------------
# Accessors
# ---------------------------------------------------------------------------

function _vertical_check_layer(g::VerticalGrid, k::Int)
    (1 ≤ k ≤ n_cells(g)) ||
        throw(BoundsError(g, k))
    return nothing
end

"""
    cell_centers(g::VerticalGrid, [k::Int]) -> Vector{T} | T

With no index, return all mid-layer coordinate values (length `nz`). With an
integer `k` (1-based), return the scalar center at that layer.
"""
cell_centers(g::VerticalGrid) = g.centers
function cell_centers(g::VerticalGrid, k::Integer)
    _vertical_check_layer(g, Int(k))
    return g.centers[k]
end

"""
    cell_widths(g::VerticalGrid, [k::Int]) -> Vector{T} | T

Layer thicknesses in native units, always positive. With no index returns
the full widths vector (length `nz`).
"""
cell_widths(g::VerticalGrid) = g.widths
function cell_widths(g::VerticalGrid, k::Integer)
    _vertical_check_layer(g, Int(k))
    return g.widths[k]
end

"""
    neighbors(g::VerticalGrid, k::Int) -> Dict{Symbol,Int}

Axis-aligned vertical neighbours of layer `k` (1-based). Returns a dict
with `:down` → `k-1` and/or `:up` → `k+1`. Bottom / top layers drop the
out-of-range side, keeping the return type concrete.
"""
function neighbors(g::VerticalGrid, k::Integer)
    _vertical_check_layer(g, Int(k))
    out = Dict{Symbol, Int}()
    if k > 1
        out[:down] = Int(k) - 1
    end
    if Int(k) < n_cells(g)
        out[:up] = Int(k) + 1
    end
    return out
end

"""
    metric_eval(g::VerticalGrid, name::Symbol, k::Int) -> T

Evaluate a named metric at layer `k` (1-based). Supported names:

- `:dz` — layer thickness (native units).
- `:z` — cell-centre value (native units).
- `:sigma` — sigma at cell centre; valid for sigma-like coordinates
  (`:sigma`, `:hybrid_sigma_theta`, `:eta`).
- `:pressure` — reference pressure at cell centre, averaged across the
  layer's two interfaces (`p = ak + bk * p0`). Requires hybrid
  coefficients.
- `:ak`, `:bk` — hybrid coefficients averaged across the layer's two
  interfaces. Requires hybrid coefficients.
"""
function metric_eval(g::VerticalGrid{T}, name::Symbol, k::Integer) where {T}
    name in _VERTICAL_METRIC_NAMES ||
        throw(ArgumentError("vertical: unknown metric :$name"))
    _vertical_check_layer(g, Int(k))
    if name === :dz
        return g.widths[k]
    elseif name === :z
        return g.centers[k]
    elseif name === :sigma
        (g.coordinate === :sigma ||
            g.coordinate === :hybrid_sigma_theta ||
            g.coordinate === :eta) ||
            throw(ArgumentError(
                "vertical: :sigma undefined for coordinate :$(g.coordinate)"
            ))
        return g.centers[k]
    elseif name === :pressure
        (length(g.ak) > 0 && length(g.bk) > 0) ||
            throw(ArgumentError(
                "vertical: :pressure requires hybrid ak/bk " *
                    "(coordinate :$(g.coordinate) has none)"
            ))
        p_lo = g.ak[k] + g.bk[k] * g.p0
        p_hi = g.ak[k + 1] + g.bk[k + 1] * g.p0
        return (p_lo + p_hi) / T(2)
    elseif name === :ak
        length(g.ak) > 0 ||
            throw(ArgumentError("vertical: :ak unavailable (no hybrid coefficients)"))
        return (g.ak[k] + g.ak[k + 1]) / T(2)
    else  # :bk
        length(g.bk) > 0 ||
            throw(ArgumentError("vertical: :bk unavailable (no hybrid coefficients)"))
        return (g.bk[k] + g.bk[k + 1]) / T(2)
    end
end

metric_eval(g::VerticalGrid, name::AbstractString, k::Integer) =
    metric_eval(g, Symbol(name), k)

# ---------------------------------------------------------------------------
# .esm lowering + provenance
# ---------------------------------------------------------------------------

function _vertical_provenance(g::VerticalGrid)
    return Dict{String, Any}(
        "binding" => "julia",
        "binding_version" => _VERTICAL_FAMILY_VERSION,
        "family" => "vertical",
        "version" => _VERTICAL_FAMILY_VERSION,
        "coordinate" => String(g.coordinate),
        "dtype" => g.dtype,
    )
end

"""
    to_esm(g::VerticalGrid) -> Dict{String,Any}

Declarative `.esm` lowering per the 2026-04-20 scope correction. The output
carries the family + coordinate kind + interface levels (+ hybrid
coefficients when applicable) + provenance; derived arrays (centres,
widths) do not appear in the wire form. Matches the Python reference
binding's key set so cross-binding byte-equality is achievable by the
conformance harness.
"""
function to_esm(g::VerticalGrid)
    options = Dict{String, Any}(
        "coordinate" => String(g.coordinate),
        "nz" => n_cells(g),
        "levels" => Float64[Float64(x) for x in g.levels],
    )
    if length(g.ak) > 0
        options["ak"] = Float64[Float64(x) for x in g.ak]
    end
    if length(g.bk) > 0
        options["bk"] = Float64[Float64(x) for x in g.bk]
    end
    if g.coordinate === :eta || g.coordinate === :hybrid_sigma_theta ||
            length(g.ak) > 0 || length(g.bk) > 0
        options["p0"] = Float64(g.p0)
    end
    return Dict{String, Any}(
        "family" => "vertical",
        "topology" => "column",
        "dtype" => g.dtype,
        "ndim" => 1,
        "ghosts" => g.ghosts,
        "n_cells" => n_cells(g),
        "n_vertices" => n_vertices(g),
        "n_edges" => n_edges(g),
        "options" => options,
        "provenance" => _vertical_provenance(g),
        "schema_version" => _VERTICAL_FAMILY_VERSION,
    )
end

family(::VerticalGrid) = "vertical"

# Public namespace registration: the `vertical` entry point in the
# `EarthSciDiscretizations.grids` submodule (defined at the bottom of
# `src/EarthSciDiscretizations.jl`) aliases `_vertical` here. See
# GRIDS_API.md §2.3 for the public call form.

File: typescript/src/grids/vertical.ts
/**
 * Vertical grid family (1D column) — accessor runtime.
 *
 * Per `docs/GRIDS_API.md` §2.6 and the 2026-04-20 mayor correction in bead
 * dsc-0k6: `toESM()` emits a small declarative config (family + coordinate
 * kind + interface levels + optional hybrid coefficients), NOT a serialized
 * geometry blob. Centers and widths are derived on demand from the
 * interface `levels` via pure arithmetic.
 *
 * Rhymes with the Python reference in
 * `python/src/earthsci_toolkit/grids/vertical.py`. Indices are 0-based.
 *
 * Supported coordinate kinds:
 *
 * | Kind                   | Value domain                  | Required options        |
 * |------------------------|-------------------------------|-------------------------|
 * | `"sigma"`              | [0, 1]; 1 = surface, 0 = top  | `nz` or `levels`        |
 * | `"eta"`                | hybrid sigma-pressure (NCAR)  | `ak`, `bk`              |
 * | `"z"`                  | geometric altitude (m)        | `levels` (increasing)   |
 * | `"theta"`              | potential temperature (K)     | `levels` (increasing)   |
 * | `"hybrid_sigma_theta"` | blended sigma→theta           | `nz` or `levels`        |
 * | `"z_star"`             | generalized height            | `levels` (increasing)   |
 */

import type { Dtype, Grid } from "./types.js";

const VERTICAL_API_VERSION = "1.0.0";
const VERTICAL_BINDING_VERSION = "0.1.0";

export type VerticalCoordinate =
  | "sigma"
  | "eta"
  | "z"
  | "theta"
  | "hybrid_sigma_theta"
  | "z_star";

const COORDINATES: ReadonlyArray<VerticalCoordinate> = [
  "sigma",
  "eta",
  "z",
  "theta",
  "hybrid_sigma_theta",
  "z_star",
];

export type VerticalMetricName =
  | "dz"
  | "z"
  | "sigma"
  | "pressure"
  | "ak"
  | "bk";

const METRIC_NAMES: ReadonlyArray<VerticalMetricName> = [
  "dz",
  "z",
  "sigma",
  "pressure",
  "ak",
  "bk",
];

export interface VerticalOpts {
  coordinate: VerticalCoordinate;
  nz?: number;
  levels?: ReadonlyArray<number>;
  ak?: ReadonlyArray<number>;
  bk?: ReadonlyArray<number>;
  p0?: number;
  transition?: number;
  dtype?: Dtype;
  ghosts?: number;
}

export interface VerticalNeighbors {
  down?: number;
  up?: number;
}

export interface VerticalGrid extends Grid {
  readonly family: "vertical";
  readonly topology: "column";
  readonly coordinate: VerticalCoordinate;
  readonly nz: number;
  readonly levels: number[];
  readonly centers: number[];
  readonly widths: number[];
  readonly ak: number[];
  readonly bk: number[];
  readonly p0: number;
  readonly transition: number | null;
  readonly ghosts: number;
  readonly ndim: 1;
  readonly n_cells: number;
  readonly n_vertices: number;
  readonly n_edges: number;

  cell_centers(k?: number): number | number[];
  cell_widths(k?: number): number | number[];
  neighbors(k: number): VerticalNeighbors;
  metric_eval(name: VerticalMetricName, k: number): number;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isNonNegativeInteger(v: unknown): v is number {
  return typeof v === "number" && Number.isInteger(v) && v >= 0;
}

function isPositiveFiniteNumber(v: unknown): v is number {
  return typeof v === "number" && Number.isFinite(v) && v > 0;
}

function coerceFiniteArray(
  values: ReadonlyArray<number>,
  label: string,
): number[] {
  const out: number[] = new Array(values.length);
  for (let i = 0; i < values.length; i++) {
    const v = values[i];
    if (typeof v !== "number" || !Number.isFinite(v)) {
      throw new RangeError(
        `vertical: ${label}[${i}] must be a finite number (got ${String(v)})`,
      );
    }
    out[i] = v;
  }
  return out;
}

function coerceLevels(
  levels: ReadonlyArray<number>,
  label: string,
  mustDecrease: boolean,
  domain: [number, number] | null,
): number[] {
  if (!Array.isArray(levels)) {
    throw new TypeError(
      `vertical: \`${label}\` must be an array of numbers (got ${typeof levels})`,
    );
  }
  if (levels.length < 2) {
    throw new RangeError(
      `vertical: \`${label}\` must have >= 2 entries (got ${levels.length})`,
    );
  }
  const arr = coerceFiniteArray(levels, label);
  if (domain !== null) {
    const [lo, hi] = domain;
    for (let i = 0; i < arr.length; i++) {
      if (arr[i] < lo || arr[i] > hi) {
        throw new RangeError(
          `vertical: \`${label}\` entries must lie in [${lo}, ${hi}] (index ${i} = ${arr[i]})`,
        );
      }
    }
  }
  for (let i = 0; i + 1 < arr.length; i++) {
    const d = arr[i + 1] - arr[i];
    if (mustDecrease ? !(d < 0) : !(d > 0)) {
      throw new RangeError(
        `vertical: \`${label}\` must be strictly ${
          mustDecrease ? "decreasing" : "increasing"
        } (violation at index ${i})`,
      );
    }
  }
  return arr;
}

function coerceHybrid(
  coeffs: ReadonlyArray<number> | undefined,
  expectedLen: number,
  label: string,
): number[] {
  if (coeffs === undefined || coeffs === null) {
    throw new TypeError(`vertical: \`${label}\` is required`);
  }
  if (!Array.isArray(coeffs)) {
    throw new TypeError(
      `vertical: \`${label}\` must be an array of numbers (got ${typeof coeffs})`,
    );
  }
  if (coeffs.length !== expectedLen) {
    throw new RangeError(
      `vertical: \`${label}\` must have length nz+1 = ${expectedLen}; got ${coeffs.length}`,
    );
  }
  return coerceFiniteArray(coeffs, label);
}

function uniformSigmaLevels(nz: number): number[] {
  const out: number[] = new Array(nz + 1);
  for (let k = 0; k <= nz; k++) out[k] = 1 - k / nz;
  return out;
}

function centersAndWidths(levels: ReadonlyArray<number>): {
  centers: number[];
  widths: number[];
} {
  const n = levels.length - 1;
  const centers: number[] = new Array(n);
  const widths: number[] = new Array(n);
  for (let k = 0; k < n; k++) {
    centers[k] = 0.5 * (levels[k] + levels[k + 1]);
    widths[k] = Math.abs(levels[k + 1] - levels[k]);
  }
  return { centers, widths };
}

function checkLayer(k: unknown, nz: number): void {
  if (typeof k !== "number" || !Number.isInteger(k)) {
    throw new TypeError(
      `vertical: k must be an integer (got ${String(k)})`,
    );
  }
  if (k < 0 || k >= nz) {
    throw new RangeError(
      `vertical: k=${k} out of range [0, ${nz})`,
    );
  }
}

// ---------------------------------------------------------------------------
// Generator
// ---------------------------------------------------------------------------

export function vertical(opts: VerticalOpts): VerticalGrid {
  if (opts === undefined || opts === null) {
    throw new TypeError("vertical: options object is required");
  }

  const coordinate = opts.coordinate;
  if (coordinate === undefined || coordinate === null) {
    throw new TypeError("vertical: required option `coordinate` is missing");
  }
  if (typeof coordinate !== "string") {
    throw new TypeError(
      `vertical: \`coordinate\` must be a string (got ${typeof coordinate})`,
    );
  }
  if (!COORDINATES.includes(coordinate as VerticalCoordinate)) {
    throw new RangeError(
      `vertical: unknown coordinate '${coordinate}'; expected one of ${COORDINATES.join(
        ", ",
      )}`,
    );
  }

  const dtype: Dtype = opts.dtype ?? "float64";
  if (dtype !== "float64" && dtype !== "float32") {
    throw new RangeError(
      `vertical: \`dtype\` must be 'float64' or 'float32' (got ${String(dtype)})`,
    );
  }

  const ghosts = opts.ghosts ?? 0;
  if (!isNonNegativeInteger(ghosts)) {
    throw new RangeError(
      `vertical: \`ghosts\` must be a non-negative integer (got ${String(ghosts)})`,
    );
  }

  const p0 = opts.p0 ?? 1.0e5;
  if (!isPositiveFiniteNumber(p0)) {
    throw new RangeError(
      `vertical: \`p0\` must be a positive finite number (got ${String(p0)})`,
    );
  }

  let nz = opts.nz;
  if (nz !== undefined) {
    if (!Number.isInteger(nz) || (nz as number) < 1) {
      throw new RangeError(
        `vertical: \`nz\` must be an integer >= 1 (got ${String(nz)})`,
      );
    }
  }

  let levels: number[];
  let ak: number[] = [];
  let bk: number[] = [];
  let transition: number | null = null;

  if (coordinate === "sigma") {
    if (opts.levels !== undefined) {
      levels = coerceLevels(opts.levels, "levels", true, [0, 1]);
      if (nz !== undefined && levels.length !== nz + 1) {
        throw new RangeError(
          `vertical: sigma: nz=${nz} inconsistent with levels length ${levels.length}`,
        );
      }
    } else {
      if (nz === undefined) {
        throw new TypeError("vertical: 'sigma' requires `nz` or `levels`");
      }
      levels = uniformSigmaLevels(nz);
    }
  } else if (
    coordinate === "z" ||
    coordinate === "theta" ||
    coordinate === "z_star"
  ) {
    if (opts.levels === undefined) {
      throw new TypeError(
        `vertical: '${coordinate}' requires explicit \`levels\``,
      );
    }
    levels = coerceLevels(opts.levels, "levels", false, null);
    if (nz !== undefined && levels.length !== nz + 1) {
      throw new RangeError(
        `vertical: ${coordinate}: nz=${nz} inconsistent with levels length ${levels.length}`,
      );
    }
  } else if (coordinate === "eta") {
    if (opts.ak === undefined) {
      throw new TypeError("vertical: 'eta' requires `ak` (length nz+1)");
    }
    if (opts.bk === undefined) {
      throw new TypeError("vertical: 'eta' requires `bk` (length nz+1)");
    }
    if (!Array.isArray(opts.ak) || !Array.isArray(opts.bk)) {
      throw new TypeError(
        "vertical: 'eta' `ak` and `bk` must be arrays of numbers",
      );
    }
    if (opts.ak.length !== opts.bk.length) {
      throw new RangeError(
        `vertical: 'eta' ak/bk must have equal length; got ${opts.ak.length} vs ${opts.bk.length}`,
      );
    }
    const nzEff = nz !== undefined ? nz : opts.ak.length - 1;
    if (nzEff < 1) {
      throw new RangeError(`vertical: nz must be >= 1 (got ${nzEff})`);
    }
    ak = coerceHybrid(opts.ak, nzEff + 1, "ak");
    bk = coerceHybrid(opts.bk, nzEff + 1, "bk");
    const sigma: number[] = new Array(ak.length);
    for (let i = 0; i < ak.length; i++) sigma[i] = ak[i] / p0 + bk[i];
    for (let i = 0; i + 1 < sigma.length; i++) {
      if (!(sigma[i + 1] - sigma[i] < 0)) {
        throw new RangeError(
          "vertical: 'eta' synthesized sigma (ak/p0 + bk) must be strictly decreasing",
        );
      }
    }
    levels = sigma;
    nz = nzEff;
  } else {
    // hybrid_sigma_theta
    if (opts.levels === undefined && nz === undefined) {
      throw new TypeError(
        "vertical: 'hybrid_sigma_theta' requires `nz` or `levels`",
      );
    }
    if (opts.levels !== undefined) {
      levels = coerceLevels(opts.levels, "levels", true, [0, 1]);
      if (nz !== undefined && levels.length !== nz + 1) {
        throw new RangeError(
          `vertical: hybrid_sigma_theta: nz=${nz} inconsistent with levels length ${levels.length}`,
        );
      }
    } else {
      levels = uniformSigmaLevels(nz as number);
    }
    if (opts.transition !== undefined) {
      if (
        typeof opts.transition !== "number" ||
        !Number.isFinite(opts.transition) ||
        !(opts.transition > 0 && opts.transition < 1)
      ) {
        throw new RangeError(
          `vertical: hybrid_sigma_theta \`transition\` must be in (0, 1); got ${String(opts.transition)}`,
        );
      }
      transition = opts.transition;
    }
    ak =
      opts.ak !== undefined ? coerceHybrid(opts.ak, levels.length, "ak") : [];
    bk =
      opts.bk !== undefined ? coerceHybrid(opts.bk, levels.length, "bk") : [];
  }

  const nzResolved = levels.length - 1;
  const { centers, widths } = centersAndWidths(levels);

  const grid: VerticalGrid = {
    family: "vertical",
    topology: "column",
    dtype,
    coordinate,
    nz: nzResolved,
    levels,
    centers,
    widths,
    ak,
    bk,
    p0,
    transition,
    ghosts,
    ndim: 1,
    n_cells: nzResolved,
    n_vertices: nzResolved + 1,
    n_edges: nzResolved,

    cell_centers(k?: number): number | number[] {
      if (k === undefined) return centers.slice();
      checkLayer(k, nzResolved);
      return centers[k];
    },

    cell_widths(k?: number): number | number[] {
      if (k === undefined) return widths.slice();
      checkLayer(k, nzResolved);
      return widths[k];
    },

    neighbors(k: number): VerticalNeighbors {
      checkLayer(k, nzResolved);
      const out: VerticalNeighbors = {};
      if (k > 0) out.down = k - 1;
      if (k < nzResolved - 1) out.up = k + 1;
      return out;
    },

    metric_eval(name: VerticalMetricName, k: number): number {
      if (!METRIC_NAMES.includes(name)) {
        throw new RangeError(`vertical: unknown metric '${String(name)}'`);
      }
      checkLayer(k, nzResolved);
      switch (name) {
        case "dz":
          return widths[k];
        case "z":
          return centers[k];
        case "sigma":
          if (
            coordinate === "sigma" ||
            coordinate === "hybrid_sigma_theta" ||
            coordinate === "eta"
          ) {
            return centers[k];
          }
          throw new RangeError(
            `vertical: 'sigma' undefined for coordinate '${coordinate}'`,
          );
        case "pressure":
          if (ak.length === 0 || bk.length === 0) {
            throw new RangeError(
              `vertical: 'pressure' requires hybrid ak/bk (coordinate '${coordinate}' has none)`,
            );
          }
          return 0.5 * (ak[k] + bk[k] * p0 + ak[k + 1] + bk[k + 1] * p0);
        case "ak":
          if (ak.length === 0) {
            throw new RangeError(
              "vertical: 'ak' unavailable (no hybrid coefficients)",
            );
          }
          return 0.5 * (ak[k] + ak[k + 1]);
        case "bk":
          if (bk.length === 0) {
            throw new RangeError(
              "vertical: 'bk' unavailable (no hybrid coefficients)",
            );
          }
          return 0.5 * (bk[k] + bk[k + 1]);
      }
    },

    toESM(): object {
      const options: Record<string, unknown> = {
        coordinate,
        nz: nzResolved,
        levels: levels.slice(),
      };
      if (ak.length > 0) options.ak = ak.slice();
      if (bk.length > 0) options.bk = bk.slice();
      if (
        coordinate === "eta" ||
        coordinate === "hybrid_sigma_theta" ||
        ak.length > 0 ||
        bk.length > 0
      ) {
        options.p0 = p0;
      }
      if (transition !== null) {
        options.transition = transition;
      }
      return {
        family: "vertical",
        topology: "column",
        dtype,
        ndim: 1,
        ghosts,
        n_cells: nzResolved,
        n_vertices: nzResolved + 1,
        n_edges: nzResolved,
        options,
        provenance: {
          binding: "typescript",
          binding_version: VERTICAL_BINDING_VERSION,
          family: "vertical",
          version: VERTICAL_API_VERSION,
          coordinate,
          dtype,
        },
        schema_version: VERTICAL_API_VERSION,
      };
    },
  };

  return grid;
}
